File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/binary_tree/P652_FindDuplicateSubtrees.java
package com.dsa.leetcode.labuladong.binary_tree;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class P652_FindDuplicateSubtrees {
    private Map<String, Integer> seen = new HashMap<>();
    private List<TreeNode> res = new ArrayList<>();

    public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
        traverse(root);
        return res;
    }

    private String traverse(TreeNode root) {
        if (root == null) {
            return "#";
        }
        String left = traverse(root.left);
        String right = traverse(root.right);
        String str = left + "," + right + "," + root.val;
        Integer count = seen.getOrDefault(str, 0);
        // 说明已经重复
        if (count == 1) {
            res.add(root);
        }
        seen.put(str, count + 1);
        return str;
    }
}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/binary_tree/P652_FindDuplicateSubtrees.py
# author: tobebetter9527
# since: 2024/3/5 8:42
from typing import Optional, List

from com.dsa.leetcode.labuladong.binary_tree import TreeNode
from collections import defaultdict


class Solution:
    def __init__(self):
        self.map = defaultdict(int)
        self.res = []

    def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
        self.traverse(root)
        return self.res

    def traverse(self, root):
        if not root:
            return "#"
        left = self.traverse(root.left)
        right = self.traverse(root.right)
        subtree = left + "," + right + "," + str(root.val)
        count = self.map[subtree]
        if count == 1:
            self.res.append(root)
        self.map[subtree] = count + 1
        return subtree
